src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+      SumAsProd _ _ _ -> unknown atom
+      LabelCon _ -> invalid "LabelCon"
+      Newtype _ e -> summary e
+      BaseTypeRef _ -> invalid "BaseTypeRef"
+      TabRef _ -> invalid "TabRef"
+      ConRef _ -> invalid "ConRef"
+      ExplicitDict _ _ -> invalid "ExplicitDict"
+      DictHole _ _ -> invalid "DictHole"
+
+unknown :: HoistableE e => e n -> OCCM n (IxExpr n)
+unknown obj = do
+  freeIxs <- mapM ixExpr $ freeAtomVarsList obj
+  let freeIterationNames = nub $ concatMap freeAtomVarsList freeIxs
+  return $ Deterministic $ freeIterationNames
+
+-- === The actual occurrence analysis ===
+
+class HasOCC (e::E) where
+  occ :: Access n -> e n -> OCCM n (e n)
+  default occ :: (GenericE e, HasOCC (RepE e)) => Access n -> e n -> OCCM n (e n)
+  occ a e = confuseGHC >>= \_ -> toE <$> occ a (fromE e)
+  {-# INLINE occ #-}
+
+instance HasOCC (Name AtomNameC) where
+  occ a n = modify (<> FV (singletonNameMapE n $ AccessInfo 1 a)) $> n
+  {-# INLINE occ #-}
+instance HasOCC (Name HandlerNameC ) where occ _ n = return n
+instance HasOCC (Name DataDefNameC ) where occ _ n = return n
+instance HasOCC (Name EffectNameC  ) where occ _ n = return n
+instance HasOCC (Name InstanceNameC) where occ _ n = return n
+instance HasOCC (Name ClassNameC   ) where occ _ n = return n
+instance HasOCC (Name SpecializedDictNameC) where occ _ n = return n
+
+instance HasOCC Block where
+  occ a (Block ann decls ans) = case (ann, decls) of
+    (NoBlockAnn      , Empty) -> Block NoBlockAnn Empty <$> occ a ans
+    (NoBlockAnn      , _    ) -> error "should be unreachable"
+    (BlockAnn ty effs, _    ) -> do
+      -- The free vars accumulated in the state of OCCM should
+      -- correspond to the free vars of the Abs of the block answer,
+      -- by the decls traversed so far. occNest takes care to uphold
+      -- this invariant, but we temporarily reset the state to an
+      -- empty map, just so that names from the surrounding block
+      -- don't end up influencing elimination decisions here. Note
+      -- that we restore the state (and accumulate free vars of the
+      -- analyzed block into it) right after occNest.
+      Abs decls' ans' <- locally $ occNest a decls ans
+      ty' <- occ accessOnce ty  -- TODO Is free variable counting enough here?
+      countFreeVarsAsOccurrences effs
+      return $ Block (BlockAnn ty' effs) decls' ans'
+
+-- TODO(optimization) Could reuse the free variable caching from dce here too.
+
+data ElimResult n where
+  ElimSuccess :: Abs (Nest Decl) Atom n -> ElimResult n
+  ElimFailure :: Decl n l -> UsageInfo -> Abs (Nest Decl) Atom l -> ElimResult n
+
+occNest :: Access n -> Nest Decl n l -> Atom l
+        -> OCCM n (Abs (Nest Decl) Atom n)
+occNest a decls ans = case decls of
+  Empty -> Abs Empty <$> occ a ans
+  Nest d@(Let _ binding) ds -> do
+    isPureDecl <- isPure binding
+    -- Note that we only ever analyze the abs below under this refreshAbs,
```

Comment:
What is "the abs" here?

Reply:
I actually don't think this comment matters either.